### PR TITLE
[TEP0138] Decouple v1beta1 beta feature validation

### PR DESCRIPTION
**Pull request by JeromeJu**
# Changes
This PR contains 2 commits. The first decouples the existing beta 
feature validation in v1beta1. Prior to this change, beta features are 
regarded as stable in v1beta1 apiVersion and they are validated only in
 v1 api behind beta enable-api-fields but not in v1beta1. This PR removes 
the gapbetween the validations of the stable features in v1 and v1beta1.

The second commit moved Task.Validate and Pipeline.Validate to
TaskSpec.Validate and PipelineSpec.Validate.

It also updates the api compatibility policy to clarify that feature
stability levels are independent on CRD apiVersions.

Fixes: https://github.com/tektoncd/pipeline/issues/6592
Part of: #7177 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
v1beta1 CRDs with beta features now requires `beta` `enable-api-fields`. More specifically, users who have been accidentally using beta features resolvers, object array params and results with `enable-api-fields` set to `stable` now needs to change to `enable-api-fields=beta`
```

### CODE REVIEW

**Comment by JeromeJu** [comment]
/hold
wait till https://github.com/tektoncd/community/pull/1034 gets merged first

File: pkg/apis/pipeline/v1beta1/pipeline_types_test.go
Comment (chitrangpatel) [inline pkg/apis/pipeline/v1beta1/pipeline_types_test.go:226]: Since this was bool earlier, the default was `false`. What is the default now? Is it an empty string? May be explicitly set `enableAPIFields` for all tests to the desired value instead of just setting to `beta` where needed?

File: pkg/reconciler/pipelinerun/resources/pipelineref.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/pipelineref.go:144]: This comment is no longer relevant and can be removed. I think we can also move this function call out of the switch statement.

File: pkg/apis/pipeline/v1beta1/taskref_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskref_validation.go:40]: can you please add tests for this change?

**Review by lbernick — COMMENTED** [review]
I think this PR should also include the changes from https://github.com/tektoncd/pipeline/pull/6995, plus updating the docs on "beta features" and "alpha features" to clarify that they don't depend on api version.

File: api_compatibility_policy.md
Comment (JeromeJu) [reply api_compatibility_policy.md:100]: Thanks! I think the columns for the previous chart was indicating the apiVersion with `v1` and `beta` (indicating v1beta1). I agree that this is confusing and adding the first column to `enable-api-fields` value does make sense!

File: pkg/apis/pipeline/v1/pipeline_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pipeline_validation.go:57]: Thanks! This makes sense, I've tried to elaborate in https://github.com/tektoncd/pipeline/issues/7077 on why we need to move it into the member function of PipelineSpec. And I am planning to separate this move out as a 2nd commit in this Pull Request for the purpose of granularity.

File: pkg/apis/pipeline/v1beta1/taskref_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/taskref_validation.go:40]: Thanks! Added and similarly for pipelineref test.

File: pkg/apis/pipeline/v1beta1/pipeline_types_test.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipeline_types_test.go:226]: Thanks that makes sense! I just change the condition to if not set just do not update the ctx configMap value, this might not require maintainers to edit every time even if it is just ctx.background. WDYT?

File: pkg/reconciler/pipelinerun/resources/pipelineref.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/pipelineref.go:143]: The analogous comment should also be removed from taskref.go

**Comment by lbernick** [comment]
please also update the chart in this docs section https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#beta-features

File: pkg/apis/pipeline/v1beta1/pipeline_types_test.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipeline_types_test.go:226]: It seems that the reason for this change here is that resolver was an alpha field and this was previously used for setting alpha fields but now it has become beta. I think just in case that we are having more alpha features being used in this case, leaving it as `enableAPIFields` might leave more future room instead of changing back to enableAlphaFields or WithContext again?

File: pkg/apis/pipeline/v1beta1/task_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/task_validation.go]: Thanks. This was off during rebase.

File: api_compatibility_policy.md
Comment (JeromeJu) [reply api_compatibility_policy.md:93]: Hmm for this part I was trying to call out api driven fields were guarded behind `enable-api-fields`. I think this would introduce ambiguity on whether other behavioural flags would also need to set `enable-api-fields` to `alpha` to get turned on? I wanted also to add some doc for behavioural flags after we introduce per feature flag but not to conflate them here. Please lemme know if that makes sense.

File: docs/additional-configs.md
Comment (JeromeJu) [reply docs/additional-configs.md]: Thanks!

**Comment by JeromeJu** [comment]
When I rerun the pull-tekton-integration test:
[Previous 3 tests + 4 examples fails](https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/6941/pull-tekton-pipeline-integration-tests/1696915876025995264) -> [1 fails example tests](https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/6941/pull-tekton-pipeline-integration-tests/1696935435529359360)

**Comment by JeromeJu** [comment]
Looks like there has been a job that succeeded from the CI job history.
/retest

**Review by lbernick — APPROVED** [review]

File: api_compatibility_policy.md
Comment (JeromeJu) [reply api_compatibility_policy.md]: Thanks, updated!

**Comment by JeromeJu** [comment]
/unhold 
[TEP0138](https://github.com/tektoncd/community/pull/1034) is merged.

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:99]: do you think it is ok to add a link to the list of our beta features here? So we can make sure the validation here covers all the features

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:3919]: what does "without the feature flag set" mean here? We will set the enable-api-fields flag when running the cases

File: pkg/apis/pipeline/v1beta1/pipelineref_validation.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1beta1/pipelineref_validation.go:38]: can this be added to `ValidateBetaFields`?

Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipelineref_validation.go:38]: I don't think that might be helpful for the codebase since that's pipelineRef rather than for the pipelineSpec?

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:99]: Thanks for the suggestion @Yongxuanzhang  
I think the point for this func is to cover all existing beta features related with the pipelineSpec that are coupled before (the beta features that had different validations in v1 versus v1beta1). So not essentially all beta features here.

I understand the point here is to give users/cluster operators full notice of the change ahead with the transition to per-feature flag.
I've put a list in https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md as a cutoff of all existing beta/alpha features that are going to stay behind `enable-api-fields` and such validations. Does that sound good to add them in the api-compatibility policy update PR with Per-feature flag introduced to give heads up since this PR mainly handles the decoupling?

Comment (Yongxuanzhang) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:99]: Oh I think I misunderstood the code here, I thought it will cover all but the pipelineref is in another function 😂 I think it's fine.

File: pkg/apis/pipeline/v1beta1/pipelineref_validation.go
Comment (Yongxuanzhang) [reply pkg/apis/pipeline/v1beta1/pipelineref_validation.go:38]: I see! Thanks!

**Review by Yongxuanzhang — APPROVED** [review]

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go]: Oh thanks for the catch! This musth be left over with some refactoring. I corrected it and opened a minor edition to test names:
PTAL: https://github.com/tektoncd/pipeline/pull/7198
🙏

**Comment by JeromeJu** [comment]
👋  @tektoncd/core-maintainers  
It looks like this PR shall be ready to merge with some approvals already. But it would be much appreciated if there are more eyes 👀  on this from you since this is going to be a breaking change to some specific beta feature users according to our discussion in [TEP0138 ](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md). 🙏  PTAL

**Comment by Yongxuanzhang** [comment]
> 👋 @tektoncd/core-maintainers It looks like this PR shall be ready to merge with some approvals already. But it would be much appreciated if there are more eyes 👀 on this since this is going to be a breaking change to some specific beta feature users according to our discussion and consensus built in [TEP0138 ](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md). 🙏
> 
> PTAL ❤️ 😸

Are we going to send emails to tekton users?

**Comment by JeromeJu** [comment]
> > 👋 @tektoncd/core-maintainers It looks like this PR shall be ready to merge with some approvals already. But it would be much appreciated if there are more eyes 👀 on this since this is going to be a breaking change to some specific beta feature users according to our discussion and consensus built in [TEP0138 ](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md). 🙏
> > PTAL ❤️ 😸
> 
> Are we going to send an email to tekton users?

Good call! We would need to send an email with the 0.53 LTS release clarifying the changes (decoupling feature&API versioning, introducing per-feature flags and the API compatibility policy updates) IIUC.

**Comment by pritidesai** [comment]
Should we indicate the feature graduation process documented here only applies to the features driven by `enable-api-fields`? Does it apply to all existing and any new feature?

https://github.com/tektoncd/pipeline/blob/3819f4f48b326dee7ce644295a0a580d5674fd20/api_compatibility_policy.md#feature-graduation-process

I feel the graduation process is a mix of `enable-api-fields` and flag per feature.

> The dedicated feature flag for this feature will change the stability level for validation to beta.

This statement is from graduation process. Are we envisioning a flag for each new feature to have `alpha/beta/stable` values? If not, what are we going validate against?

**Comment by JeromeJu** [comment]
Thanks for the review and the checking @pritidesai 

> I feel the graduation process is a mix of `enable-api-fields` and flag per feature.
> > The dedicated feature flag for this feature will change the stability level for validation to beta.

Thanks for pointing this out. I think `validation for` is misleading here. This won't apply to per-feature flag since we don't need to validate the stability of the feature being used by then. I've updated the wording here.

> Should we indicate the feature graduation process documented here only applies to the features driven by `enable-api-fields`? Does it apply to all existing and any new feature?
> I feel the graduation process is a mix of `enable-api-fields` and flag per feature.

The graduation process shall aim to be general, not to be specifically applied to API-driven fields or behavioural flags. It is some progress that all existing and new features in pipeline should go through IIUC. (like what we have as of today, a mix of behavioural flags and API-driven features in [beta features table in addtional-configs](https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#beta-features).

File: pkg/apis/pipeline/v1/pipeline_validation.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1/pipeline_validation.go:89]: NIT: should we consider this moving to the top as a first check? not that it stops any further validation but the error regarding the feature flag misconfiguration is up front.

File: pkg/apis/pipeline/v1/task_validation.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1/task_validation.go:96]: NIT: similar concern - comment ⬆️

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:97]: NIT: this comment is little confusing, suggest:

```
ValidateBetaFields returns an error if the PipelineSpec uses beta specifications governed by `enable-api-fields` but does not have "enable-api-fields" set to "alpha" or "beta".
```

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:126]: s/tasks/finally/

File: pkg/apis/pipeline/v1/pipeline_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pipeline_validation.go]: Thanks for your time and the review! I've updated the PR according to the pointers 🙏

File: pkg/apis/pipeline/v1/pipeline_validation.go
Comment (afrittoli) [inline pkg/apis/pipeline/v1/pipeline_validation.go:67]: It sounds good - any special reason for that? 
We might have beta fields in the status as well, but those are produced by the controller, so I guess it's fine to skip validation.

File: api_compatibility_policy.md
Comment (afrittoli) [inline api_compatibility_policy.md:132]: These two sentences
> GA/Stable features are enabled by default.

and 

> GA/Stable features are features that have been promoted from beta to the highest level of stability. They cannot be disabled in any CRD version.

kind of contradict each other. "Enabled by default" implies that they can be disabled, but the default is "enabled".

I would say that stable features are not controlled by the `enable-api-fields` flag anymore.
I guess that in this context we are only talking about feature flags for API-driven features - other features may still be controlled by a separate dedicated feature flag even once they're stable.

File: pkg/apis/pipeline/v1/pipelinerun_validation.go
Comment (afrittoli) [inline pkg/apis/pipeline/v1/pipelinerun_validation.go:73]: Why is this removed?

File: pkg/apis/pipeline/v1/pipeline_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pipeline_validation.go:67]: Thanks for question Andrea. Related with the comment at https://github.com/tektoncd/pipeline/pull/6941#discussion_r1364619411 (many thanks to the suggestion from Priti), I think this was moved to the front for handling validations of easier logics first. And yes I do agree that validation for the PipelineSpec shall be sufficient for the webhook at the moment:)

File: pkg/apis/pipeline/v1/pipelinerun_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pipelinerun_validation.go:73]: Exactly, and I tried to include this in the commit message. Hope this captures it.

**Comment by JeromeJu** [comment]
I've updated the PR and seems all current comments shall be resolved. Shall we merge this cc @afrittoli and @pritidesai ? 🙏

**Review by chitrangpatel — APPROVED** [review]

**Comment by jerop** [comment]
/hold

want to make sure a non-googler takes a look (feel free to unhold)

cc @pritidesai @afrittoli

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go]: I've updated it to finally, but as https://github.com/tektoncd/pipeline/issues/7258 tracks it actually does not work as expected. Will revisit when #7258 is fixed.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go]: Thanks!

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:118]: probably a nit.

Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:118]: Yeah I also had similar question when this was added for v1 at https://github.com/tektoncd/pipeline/blob/8123d6332245008983aabe1dc910a63bd2c6f261/pkg/apis/pipeline/v1/pipeline_validation.go#L119 via https://github.com/tektoncd/pipeline/commit/1535587f5b45f4e228d1e2cb941379faef0fff20#diff-7ce49bf4f55e7af22bd0780d7f11c97c4fad98e7b6e0939fad1e81fc9e8e18fc. I think the reason that this was for more clarification of String Results are stable and other 2 types are beta features.

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:127]: @JeromeJu I am sorry to go back on this but we had two new promotions - matrix and task level resources:

https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#beta-features

Do we need them here as well? or I think the features themselves have the check, is that sufficient?

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:1561]: NIT: please add doc string explaining what is being tested

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:4122]: NIT: please add doc string what is being tested

**Comment by pritidesai** [comment]
thanks @JeromeJu, the changes look good to me, I have left a few minor doc string comments in general and a question on recently promoted features.

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:127]: I think this shall be sufficient. [Task-level resources beta promotion](https://github.com/tektoncd/pipeline/pull/7225/files#diff-da7c4e39bb1d27ceff9fb7f3fbe0c355525c2925bffc965ccb91adfc5389faf3) and [Matrix beta promotion](https://github.com/tektoncd/pipeline/pull/7164#discussion_r1342768960) shall already have the beta validations covered. 

And this function was meant to bring back the validations that were once missing at the stage TEP0138 was not in place and while feature & API versioning were coupled.

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go
Comment (JeromeJu) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:1564]: Thanks Priti. Added:)

**Comment by JeromeJu** [comment]
> thanks @JeromeJu, the changes look good to me, I have left a few minor doc string comments in general and a question on recently promoted features.

Thanks @pritidesai . I've addressed the comments. Please lemme know if there are any other concerns? Or would you mind approving this given the release cutoff dates tmr 🙏

**Comment by pritidesai** [comment]
thanks @JeromeJu for addressing all the comments 🙏 

/lgtm

**Comment by pritidesai** [comment]
> /hold wait till [tektoncd/community#1034](https://github.com/tektoncd/community/pull/1034) gets merged first TBD on #6948

is there anything pending here? I think the required dependencies are addressed, please unhold this if there are no additional concerns.

**Comment by JeromeJu** [comment]
Thanks for the time reviewing and the instructive comments ! @pritidesai 

/unhold

**Pull request merged**
